python/ray/train/_internal/storage.py

```diff
@@ -317,6 +320,14 @@ class StorageContext:
     For example, on the driver, the storage context is initialized, only knowing
     the experiment path. On the Trainable actor, the trial_dir_name is accessible.
 
+    There are 3 path concepts:
+    1. *_path: A path or URI. This is the same format as the `storage_path` specified
+        by the user. This may need to be URI-joined.
+    2. *_fs_path: A path relative to the storage filesystem. This is a regular path
```

Comment:
Hmm I suspect this will get pretty confusing over time. How about using using Python types to distinguish between these?

Btw, why do we need to distinguish between (1) and (2) internally? It seems to me that (1) only appears in the public API, but internally we could immediately resolve it to (2)?

E.g., (storage_path="s3://foo/bar", fs=None) -> (S3FileSystem, "foo/bar")
(storage_path="foo/bar", fs=S3FileSystem) -> (S3FileSystem, "foo/bar")

Reply:
There's no need to distinguish between 1 and 2 internally actually. I just needed to keep 1 around to keep feature parity for some of our output `Result` APIs:

```
result_grid.experiment_path  # expects s3://
result.path  # expects s3://...
```

Reply:
I see, then we don't need (1) as a concept right? We could join the originally provided `storage_path` with the "fs_path" used internally?

Reply:
I think there are some cases where that doesn't work:

```python
fs, fs_path = fs.FileSystem.from_uri("mock://a/b/c")
# I couldn't test hdfs locally, but I'm pretty sure it's the same
# fs, fs_path = fs.FileSystem.from_uri("hdfs://a/b/c")
fs_path == "b/c"  # a is the netloc, and is captured in fs

# Incorrect full URI
result.path == "hdfs://b/c/exp_name/trial_name"
```

Reply:
What about this: we could compute the (storage_prefix, path) up-front (using as much custom logic as we want, so):

("mock://a/b/c", None) -> (MockFS, storage_prefix="mock://a", path="b/c")
("a/b/c", MockFS) -> (MockFS, storage_prefix="a", path="b/c")

Or something like this? So basically, instead of having a fs_path vs path concept, have a path and prefix concept which is a bit easier to understand.

Reply:
Ah ok, I like that idea. Some usage examples:

```
# storage.experiment_path, storage.trial_path, storage.checkpoint_path are all relative to the storage_prefix

pyarrow.fs.copy_files(source=local_dir, dest=storage.experiment_path, dest_fs=storage.storage_filesystem)

result.path = str(URI(storage.storage_prefix) / storage.trial_path)
```

There is one weird case for query parameters in a uri, which is not even a prefix:

```python
("s3://a/b/c?endpoint_override=asdf", None) -> (S3FS, storage_prefix="s3://?endpoint_override=asdf", path="a/b/c")
```

The logic to strip the path from within this uri could be a bit nasty. People have brought up this uri param thing in the past, which is why I am considering it. They could probably set these things by passing in a custom filesystem -- but there may be certain url params that aren't exposed in fs APIs.

Reply:
Sounds pretty good. I guess what we really need is the full URI parsing then? So not just prefix, but also params, etc. that we can re-assemble on the output APIs?

Reply:
Yep I think so

Reply:
I kept the `fs_` prefix so that there can still be a distinction between the user provided `storage_path` and `storage_fs_path`. 

But, looks better now that there's only 2 path types: `*_fs_path` vs. `*_local_path`. PTAL!